File: android-core/src/test/java/com/mparticle/internal/ApplicationContextWrapperTest.java
package com.mparticle.internal;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;

import com.mparticle.MParticle;
import com.mparticle.MockMParticle;
import com.mparticle.internal.ApplicationContextWrapper.ActivityLifecycleCallbackRecorder;
import com.mparticle.internal.ApplicationContextWrapper.LifeCycleEvent;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.lang.ref.WeakReference;
import java.util.LinkedList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

/**
 * Created by wpassidomo on 2/21/17.
 */

/**
 *  Test that if you add a bunch on mock activiies that it will play back in the same order.
 *  Test that it won't crash if there aren't anuy activiies added.
 *  Test that it will max out at 10.
 *  Test that it will take wierd, out of order combinations.
 *  Test that it will ONLY replay activies that match current activity.
 *  Test that if you call stopRecordLifecycles, you will not record anymore lifecycles.
 */

public class ApplicationContextWrapperTest {
    MParticle instance;
    Activity activity1 = mock(Activity.class);
    Activity activity2 = mock(Activity.class);
    Activity activity3 = mock(Activity.class);
    ActivityLifecycleCallbackRecorder activityLifecycleCallbackRecorder;
    MockApplicationContextWrapper applicationContextWrapper = new MockApplicationContextWrapper(mock(Application.class));
    WeakReference<Activity> activity1Ref = new WeakReference<Activity>(activity1);
    WeakReference<Activity> activity2Ref = new WeakReference<Activity>(activity2);
    WeakReference<Activity> activity3Ref = new WeakReference<Activity>(activity3);

    Bundle bundle1 = mock(Bundle.class);
    Bundle bundle2 = mock(Bundle.class);

    public class MockApplicationContextWrapper extends ApplicationContextWrapper {
        MockApplicationContextWrapper(Application application) {
            super(application);
        }

        @Override
        protected void attachBaseContext(Context base) {}
    }

    @Before
    public void setup() {
        activityLifecycleCallbackRecorder = applicationContextWrapper.getActivityLifecycleCallbackRecorderInstance();
        instance = new MockMParticle();
        MParticle.setInstance(instance);
    }

    private ActivityLifecycleCallbackRecorder recorder;
    private ActivityLifecycleCallbackRecordTester tester;

    @Test
    public void testRegisterActivityLifecycleCallbacksActivity1() {
        Mockito.when(instance.Internal().getKitManager().getCurrentActivity()).thenReturn(activity1Ref);
        recorder = getMixedActivityCallbacks();
        tester = new ActivityLifecycleCallbackRecordTester();
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder);
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester);
        assertListEquals(tester.getRecordedLifecycleList(), getMixedTestListActivity1());
    }

    @Test
    public void testRegisterActivityLifecycleCallbacksActivity2() {
        Mockito.when(instance.Internal().getKitManager().getCurrentActivity()).thenReturn(activity2Ref);
        recorder = getMixedActivityCallbacks();
        tester = new ActivityLifecycleCallbackRecordTester();
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder);
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester);
        assertListEquals(tester.getRecordedLifecycleList(), getMixedTestListActivity2());
    }

    @Test
    public void testRegisterActivityLifecycleCallbacksActivity3() {
        Mockito.when(instance.Internal().getKitManager().getCurrentActivity()).thenReturn(activity3Ref);
        recorder = getMixedActivityCallbacks();
        tester = new ActivityLifecycleCallbackRecordTester();
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder);
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester);
        assertListEquals(tester.getRecordedLifecycleList(), new LinkedList());
    }

    @Test
    public void testRegisterActivityLifecycleCallbacksEmpty() {
        Mockito.when(instance.Internal().getKitManager().getCurrentActivity()).thenReturn(activity1Ref);
        recorder = getEmptyActivityCallbacks();
        tester = new ActivityLifecycleCallbackRecordTester();
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder);
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester);
        assertListEquals(tester.getRecordedLifecycleList(), new LinkedList());
    }

    @Test
    public void testStopReplaying() {
        Mockito.when(instance.Internal().getKitManager().getCurrentActivity()).thenReturn(activity1Ref);
        recorder = getMixedActivityCallbacks();
        tester = new ActivityLifecycleCallbackRecordTester();
        applicationContextWrapper.setReplayActivityLifecycle(false);
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder);
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester);
        assertListEquals(tester.getRecordedLifecycleList(), new LinkedList());
        assertTrue(recorder.lifeCycleEvents.size() > 0);
    }

    private ActivityLifecycleCallbackRecorder getEmptyActivityCallbacks() {
        return applicationContextWrapper.getActivityLifecycleCallbackRecorderInstance();
    }

    private ActivityLifecycleCallbackRecorder getMixedActivityCallbacks() {
        activityLifecycleCallbackRecorder = applicationContextWrapper.getActivityLifecycleCallbackRecorderInstance();
        activityLifecycleCallbackRecorder.onActivityCreated(activity1, bundle1);
        activityLifecycleCallbackRecorder.onActivityStarted(activity1);
        activityLifecycleCallbackRecorder.onActivityResumed(activity1);
        activityLifecycleCallbackRecorder.onActivityPaused(activity1);
        activityLifecycleCallbackRecorder.onActivityCreated(activity2, bundle2);
        activityLifecycleCallbackRecorder.onActivitySaveInstanceState(activity1, null);
        activityLifecycleCallbackRecorder.onActivityStopped(activity1);
        activityLifecycleCallbackRecorder.onActivityStarted(activity2);
        activityLifecycleCallbackRecorder.onActivityResumed(activity2);
        activityLifecycleCallbackRecorder.onActivityDestroyed(activity1);
        return activityLifecycleCallbackRecorder;
    }

    private LinkedList<LifeCycleEvent> getMixedTestListActivity2() {
        LinkedList<LifeCycleEvent> testList = new LinkedList<LifeCycleEvent>();
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_CREATED, new WeakReference<Activity>(activity2), bundle2));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_STARTED, new WeakReference<Activity>(activity2)));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_RESUMED, new WeakReference<Activity>(activity2)));
        return testList;
    }

    private LinkedList<LifeCycleEvent> getMixedTestListActivity1() {
        LinkedList<LifeCycleEvent> testList = new LinkedList<LifeCycleEvent>();
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_CREATED, new WeakReference<Activity>(activity1), bundle1));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_STARTED, new WeakReference<Activity>(activity1)));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_RESUMED, new WeakReference<Activity>(activity1)));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_PAUSED, new WeakReference<Activity>(activity1)));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_SAVE_INSTANCE_STATE,new WeakReference<Activity>(activity1), null));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_STOPPED, new WeakReference<Activity>(activity1)));
        testList.addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_DESTROYED, new WeakReference<Activity>(activity1)));
        return testList;
    }

    private void assertListEquals(List list1, List list2) {
        assertEquals(list1.size(), list2.size());
        for (int i = 0; i < list1.size(); i++) {
            assertTrue(list1.get(i).equals(list2.get(i)));
        }
    }

    @TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    class ActivityLifecycleCallbackRecordTester implements Application.ActivityLifecycleCallbacks {
        LinkedList<LifeCycleEvent> lifeCycleEvents = new LinkedList<LifeCycleEvent>();
        int MAX_LIST_SIZE = 10;

        @Override
        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_CREATED, new WeakReference<Activity>(activity), savedInstanceState));
        }

        @Override
        public void onActivityStarted(Activity activity) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_STARTED, new WeakReference<Activity>(activity)));
        }

        @Override
        public void onActivityResumed(Activity activity) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_RESUMED, new WeakReference<Activity>(activity)));
        }

        @Override
        public void onActivityPaused(Activity activity) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_PAUSED, new WeakReference<Activity>(activity)));
        }

        @Override
        public void onActivityStopped(Activity activity) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_STOPPED, new WeakReference<Activity>(activity)));
        }

        @Override
        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_SAVE_INSTANCE_STATE, new WeakReference<Activity>(activity), outState));
        }

        @Override
        public void onActivityDestroyed(Activity activity) {
            getRecordedLifecycleList().addLast(applicationContextWrapper.getLifeCycleEventInstance(ApplicationContextWrapper.MethodType.ON_DESTROYED, new WeakReference<Activity>(activity)));
        }

        private LinkedList<LifeCycleEvent> getRecordedLifecycleList() {
            if (lifeCycleEvents.size() > MAX_LIST_SIZE) {
                lifeCycleEvents.removeFirst();
                return getRecordedLifecycleList();
            }
            return lifeCycleEvents;
        }
    }
}

File: android-core/src/test/kotlin/com/mparticle/internal/ApplicationContextWrapperTest.kt
package com.mparticle.internal

import android.annotation.TargetApi
import android.app.Activity
import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.content.Context
import android.os.Build
import android.os.Bundle
import com.mparticle.MParticle
import com.mparticle.MockMParticle
import com.mparticle.internal.ApplicationContextWrapper.ActivityLifecycleCallbackRecorder
import com.mparticle.internal.ApplicationContextWrapper.LifeCycleEvent
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import java.lang.ref.WeakReference
import java.util.LinkedList

/**
 * Created by wpassidomo on 2/21/17.
 */
/**
 * Test that if you add a bunch on mock activiies that it will play back in the same order.
 * Test that it won't crash if there aren't anuy activiies added.
 * Test that it will max out at 10.
 * Test that it will take weird, out of order combinations.
 * Test that it will ONLY replay activies that match current activity.
 * Test that if you call stopRecordLifecycles, you will not record anymore lifecycles.
 */
class ApplicationContextWrapperTest {
    private lateinit var instance: MParticle
    var activity1 = Mockito.mock(Activity::class.java)
    var activity2 = Mockito.mock(Activity::class.java)
    var activity3 = Mockito.mock(Activity::class.java)
    private lateinit var activityLifecycleCallbackRecorder: ActivityLifecycleCallbackRecorder
    var applicationContextWrapper = MockApplicationContextWrapper(
        Mockito.mock(
            Application::class.java
        )
    )
    var activity1Ref = WeakReference(activity1)
    var activity2Ref = WeakReference(activity2)
    var activity3Ref = WeakReference(activity3)
    var bundle1 = Mockito.mock(Bundle::class.java)
    var bundle2 = Mockito.mock(Bundle::class.java)

    inner class MockApplicationContextWrapper internal constructor(application: Application?) :
        ApplicationContextWrapper(application!!) {
        override fun attachBaseContext(base: Context) {}
    }

    @Before
    fun setup() {
        activityLifecycleCallbackRecorder =
            applicationContextWrapper.activityLifecycleCallbackRecorderInstance
        instance = MockMParticle()
        MParticle.setInstance(instance)
    }

    private lateinit var recorder: ActivityLifecycleCallbackRecorder
    private lateinit var tester: ActivityLifecycleCallbackRecordTester

    @Test
    fun testRegisterActivityLifecycleCallbacksActivity1() {
        Mockito.`when`(instance.Internal().kitManager.currentActivity).thenReturn(activity1Ref)
        recorder = mixedActivityCallbacks
        tester = ActivityLifecycleCallbackRecordTester()
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder)
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester, true)
        assertListEquals(tester.recordedLifecycleList, mixedTestListActivity1)
    }

    @Test
    fun testRegisterActivityLifecycleCallbacksActivity2() {
        Mockito.`when`(instance.Internal().kitManager.currentActivity).thenReturn(activity2Ref)
        recorder = mixedActivityCallbacks
        tester = ActivityLifecycleCallbackRecordTester()
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder)
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester, true)
        assertListEquals(tester.recordedLifecycleList, mixedTestListActivity2)
    }

    @Test
    fun testRegisterActivityLifecycleCallbacksActivity3() {
        Mockito.`when`(instance.Internal().kitManager.currentActivity).thenReturn(activity3Ref)
        recorder = mixedActivityCallbacks
        tester = ActivityLifecycleCallbackRecordTester()
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder)
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester, true)
        assertListEquals(tester.recordedLifecycleList, LinkedList<Any?>())
    }

    @Test
    fun testRegisterActivityLifecycleCallbacksEmpty() {
        Mockito.`when`(instance.Internal().kitManager.currentActivity).thenReturn(activity1Ref)
        recorder = emptyActivityCallbacks
        tester = ActivityLifecycleCallbackRecordTester()
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder)
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester, true)
        assertListEquals(tester.recordedLifecycleList, LinkedList<Any?>())
    }

    @Test
    fun testStopReplaying() {
        Mockito.`when`(instance.Internal().kitManager.currentActivity).thenReturn(activity1Ref)
        recorder = mixedActivityCallbacks
        tester = ActivityLifecycleCallbackRecordTester()
        applicationContextWrapper.isReplayActivityLifecycle = false
        applicationContextWrapper.setActivityLifecycleCallbackRecorder(recorder)
        applicationContextWrapper.registerActivityLifecycleCallbacks(tester, true)
        assertListEquals(tester.recordedLifecycleList, LinkedList<Any?>())
        Assert.assertTrue(recorder.lifeCycleEvents.size > 0)
    }

    private val emptyActivityCallbacks: ActivityLifecycleCallbackRecorder
        get() = applicationContextWrapper.activityLifecycleCallbackRecorderInstance
    private val mixedActivityCallbacks: ActivityLifecycleCallbackRecorder
        get() {
            activityLifecycleCallbackRecorder =
                applicationContextWrapper.activityLifecycleCallbackRecorderInstance
            activityLifecycleCallbackRecorder.onActivityCreated(activity1, bundle1)
            activityLifecycleCallbackRecorder.onActivityStarted(activity1)
            activityLifecycleCallbackRecorder.onActivityResumed(activity1)
            activityLifecycleCallbackRecorder.onActivityPaused(activity1)
            activityLifecycleCallbackRecorder.onActivityCreated(activity2, bundle2)
            activityLifecycleCallbackRecorder.onActivitySaveInstanceState(activity1, bundle1)
            activityLifecycleCallbackRecorder.onActivityStopped(activity1)
            activityLifecycleCallbackRecorder.onActivityStarted(activity2)
            activityLifecycleCallbackRecorder.onActivityResumed(activity2)
            activityLifecycleCallbackRecorder.onActivityDestroyed(activity1)
            return activityLifecycleCallbackRecorder
        }
    private val mixedTestListActivity2: LinkedList<LifeCycleEvent>
        get() {
            val testList = LinkedList<LifeCycleEvent>()
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_CREATED,
                    WeakReference(activity2),
                    bundle2
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_STARTED,
                    WeakReference(activity2)
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_RESUMED,
                    WeakReference(activity2)
                )
            )
            return testList
        }
    private val mixedTestListActivity1: LinkedList<LifeCycleEvent>
        get() {
            val testList = LinkedList<LifeCycleEvent>()
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_CREATED,
                    WeakReference(activity1),
                    bundle1
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_STARTED,
                    WeakReference(activity1)
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_RESUMED,
                    WeakReference(activity1)
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_PAUSED,
                    WeakReference(activity1)
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_SAVE_INSTANCE_STATE,
                    WeakReference(activity1),
                    bundle1
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_STOPPED,
                    WeakReference(activity1)
                )
            )
            testList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_DESTROYED,
                    WeakReference(activity1)
                )
            )
            return testList
        }

    private fun assertListEquals(list1: List<Any?>, list2: List<Any?>) {
        Assert.assertEquals(list1.size.toLong(), list2.size.toLong())
        for (i in list1.indices) {
            Assert.assertTrue(list1[i] == list2[i])
        }
    }

    @TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    internal inner class ActivityLifecycleCallbackRecordTester : ActivityLifecycleCallbacks {
        var lifeCycleEvents = LinkedList<LifeCycleEvent>()
        var MAX_LIST_SIZE = 10
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_CREATED,
                    WeakReference(activity),
                    savedInstanceState
                )
            )
        }

        override fun onActivityStarted(activity: Activity) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_STARTED,
                    WeakReference(activity)
                )
            )
        }

        override fun onActivityResumed(activity: Activity) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_RESUMED,
                    WeakReference(activity)
                )
            )
        }

        override fun onActivityPaused(activity: Activity) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_PAUSED,
                    WeakReference(activity)
                )
            )
        }

        override fun onActivityStopped(activity: Activity) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_STOPPED,
                    WeakReference(activity)
                )
            )
        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_SAVE_INSTANCE_STATE,
                    WeakReference(activity),
                    outState
                )
            )
        }

        override fun onActivityDestroyed(activity: Activity) {
            recordedLifecycleList.addLast(
                applicationContextWrapper.getLifeCycleEventInstance(
                    ApplicationContextWrapper.MethodType.ON_DESTROYED,
                    WeakReference(activity)
                )
            )
        }

        val recordedLifecycleList: LinkedList<LifeCycleEvent>
            get() {
                if (lifeCycleEvents.size > MAX_LIST_SIZE) {
                    lifeCycleEvents.removeFirst()
                    return recordedLifecycleList
                }
                return lifeCycleEvents
            }
    }
}
